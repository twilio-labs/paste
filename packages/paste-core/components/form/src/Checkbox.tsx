import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUID} from 'react-uid';
import {styled} from '@twilio-paste/styling-library';
import {Box} from '@twilio-paste/box';
import {BackgroundColorOptions, SpaceOptions} from '@twilio-paste/style-props';
import {
  BaseRadioCheckboxControl,
  BaseRadioCheckboxLabel,
  BaseRadioCheckboxLabelText,
  BaseRadioCheckboxHelpText,
} from './shared/BaseRadioCheckbox';

/* eslint-disable emotion/syntax-preference */
const HiddenCheckbox = styled.input({
  border: '0',
  clip: 'rect(0 0 0 0)',
  height: '1px',
  margin: '-1px',
  overflow: 'hidden',
  padding: '0',
  position: 'absolute',
  textTransform: 'none',
  whiteSpace: 'nowrap',
  width: '1px',
});
/* eslint-enable */

export interface CheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
  children: NonNullable<React.ReactNode>;
  hasError?: boolean;
  helpText?: string | React.ReactNode;
  id: string;
  indeterminate?: boolean;
  isSelectAll?: boolean;
  isSelectAllChild?: boolean;
}
const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {id, checked, children, helpText, disabled, hasError, indeterminate, isSelectAllChild, isSelectAll, ...props},
    ref
  ) => {
    const helpTextId = useUID();

    let checkboxBackground: BackgroundColorOptions | null = null;
    if (isSelectAll) checkboxBackground = 'colorBackground';
    if ((isSelectAll && checked) || (isSelectAll && indeterminate))
      checkboxBackground = 'colorBackgroundPrimaryLightest';

    let paddingLeft: SpaceOptions | null = isSelectAll ? 'space20' : null;
    if (isSelectAllChild) paddingLeft = 'space30';

    return (
      <Box
        backgroundColor={checkboxBackground}
        borderRadius={isSelectAll ? 'borderRadius10' : null}
        display="inline-flex"
        flexDirection="column"
        padding={isSelectAll ? 'space30' : null}
        paddingBottom={isSelectAll ? 'space20' : null}
        paddingLeft={paddingLeft}
        paddingRight={isSelectAllChild ? 'space30' : null}
        paddingTop={isSelectAll ? 'space20' : null}
      >
        <HiddenCheckbox
          {...props}
          aria-describedby={helpTextId}
          aria-checked={indeterminate ? 'mixed' : checked}
          aria-invalid={hasError}
          checked={checked}
          disabled={disabled}
          id={id}
          ref={ref}
          type="checkbox"
        />
        <BaseRadioCheckboxLabel disabled={disabled} htmlFor={id}>
          <BaseRadioCheckboxControl
            alignItems="center"
            borderRadius="borderRadius10"
            disabled={disabled}
            display="flex"
            position="relative"
            type="checkbox"
          >
            <Box
              alignItems="center"
              as="span"
              display="flex"
              height={!indeterminate ? 'sizeSquare25' : undefined}
              position="absolute"
              top={!indeterminate ? '3px' : undefined}
              transform={!indeterminate ? 'rotate(-45deg)' : undefined}
              width="sizeSquare30"
              _before={
                !indeterminate
                  ? {
                      borderColor: disabled && checked ? 'colorBorderDark' : 'colorBorderInverse',
                      borderRadius: 'borderRadius10',
                      borderStyle: 'solid',
                      borderWidth: 'borderWidth10',
                      bottom: '0',
                      content: '""',
                      display: 'block',
                      height: 'sizeSquare25',
                      position: 'absolute',
                      width: 'size0',
                    }
                  : undefined
              }
              _after={{
                borderColor: disabled && checked ? 'colorBorderDark' : 'colorBorderInverse',
                borderRadius: 'borderRadius10',
                borderStyle: 'solid',
                borderWidth: 'borderWidth10',
                bottom: '0',
                content: '""',
                display: 'block',
                height: 'size0',
                position: !indeterminate ? 'absolute' : undefined,
                width: 'sizeSquare30',
              }}
            />
          </BaseRadioCheckboxControl>
          <BaseRadioCheckboxLabelText fontWeight={isSelectAll ? null : 'fontWeightNormal'}>
            {children}
          </BaseRadioCheckboxLabelText>
        </BaseRadioCheckboxLabel>
        {helpText && <BaseRadioCheckboxHelpText helpTextId={helpTextId}>{helpText}</BaseRadioCheckboxHelpText>}
      </Box>
    );
  }
);
Checkbox.displayName = 'Checkbox';

if (process.env.NODE_ENV === 'development') {
  Checkbox.propTypes = {
    children: PropTypes.node.isRequired,
    hasError: PropTypes.bool,
    helpText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    id: PropTypes.string.isRequired,
    indeterminate: PropTypes.bool,
    isSelectAll: PropTypes.bool,
    isSelectAllChild: PropTypes.bool,
    value: PropTypes.string.isRequired,
  };
}

export {Checkbox};
