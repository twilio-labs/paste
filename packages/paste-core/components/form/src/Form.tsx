import { Box, safelySpreadBoxProps } from "@twilio-paste/box";
import type { BoxProps, BoxStyleProps } from "@twilio-paste/box";
import * as React from "react";

export interface FormProps extends Omit<React.ComponentPropsWithRef<"form">, "children"> {
  element?: BoxProps["element"];
  children: React.ReactNode;
  maxWidth?: BoxStyleProps["maxWidth"];
}

export const Form = React.forwardRef<HTMLDivElement, FormProps>(
  ({ element = "FORM", maxWidth, children, ...props }, ref) => (
    <Box
      as="form"
      ref={ref}
      element={element}
      maxWidth={maxWidth}
      display="flex"
      flexDirection="column"
      rowGap="space80"
      {...safelySpreadBoxProps(props)}
    >
      {children}
    </Box>
  ),
);

Form.displayName = "Form";
