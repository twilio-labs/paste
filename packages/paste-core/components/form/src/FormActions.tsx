import * as React from 'react';
import PropTypes from 'prop-types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';

export interface FormActionsProps extends Omit<React.ComponentPropsWithRef<'div'>, 'children'> {
  element?: BoxProps['element'];
  children: React.ReactNode;
}

export const FormActions = React.forwardRef<HTMLDivElement, FormActionsProps>(
  ({children, element = 'FORM_ACTIONS', ...props}, ref) => (
    <Box
      ref={ref}
      element={element}
      {...safelySpreadBoxProps(props)}
      display="flex"
      flexDirection="row"
      columnGap="space40"
      marginTop="space60"
    >
      {children}
    </Box>
  )
);

FormActions.displayName = 'FormActions';

FormActions.propTypes = {
  children: PropTypes.node,
  element: PropTypes.string,
};
