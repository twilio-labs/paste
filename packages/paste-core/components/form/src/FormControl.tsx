import * as React from 'react';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';

export interface FormControlProps extends Omit<React.ComponentPropsWithRef<'div'>, 'children'> {
  element?: BoxProps['element'];
  children: React.ReactNode;
}

export const FormControl = React.forwardRef<HTMLDivElement, FormControlProps>(
  ({children, element = 'FORM_CONTROL', ...props}, ref) => (
    <Box ref={ref} flexGrow={1} element={element} {...safelySpreadBoxProps(props)}>
      {children}
    </Box>
  )
);

FormControl.displayName = 'FormControl';
