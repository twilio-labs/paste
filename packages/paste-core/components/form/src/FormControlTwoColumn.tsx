import { Box, safelySpreadBoxProps } from "@twilio-paste/box";
import type { BoxProps } from "@twilio-paste/box";
import type { HTMLPasteProps } from "@twilio-paste/types";
import * as React from "react";

export interface FormControlTwoColumnProps extends HTMLPasteProps<"legend"> {
  /**
   * Overrides the default element name to apply unique styles with the Customization Provider
   *
   * @default 'FORM_CONTROL_TWO_COLUMN'
   * @type {BoxProps['element']}
   * @memberof FormControlTwoColumnProps
   */
  element?: BoxProps["element"];
  children: React.ReactNode;
}

export const FormControlTwoColumn = React.forwardRef<HTMLLegendElement, FormControlTwoColumnProps>(
  ({ children, element = "FORM_CONTROL_TWO_COLUMN", ...props }, ref) => (
    <Box
      display="grid"
      gridTemplateColumns="1fr 1fr"
      columnGap="space50"
      ref={ref}
      element={element}
      {...safelySpreadBoxProps(props)}
    >
      {children}
    </Box>
  ),
);

FormControlTwoColumn.displayName = "FormControlTwoColumn";
