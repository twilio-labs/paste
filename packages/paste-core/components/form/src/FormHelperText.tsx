import * as React from 'react';
import * as PropTypes from 'prop-types';
import {ValueOf} from '@twilio-paste/types';
import {Flex} from '@twilio-paste/flex';
import {Text, safelySpreadTextProps} from '@twilio-paste/text';
import {ErrorIcon} from '@twilio-paste/icons/esm/ErrorIcon';

export const FormHelperTextVariants = {
  ERROR: 'error',
} as const;

export type FormHelperTextVariants = ValueOf<typeof FormHelperTextVariants>;

export interface FormHelperTextProps extends React.HTMLAttributes<''> {
  marginTop?: 'space0';
  className?: never;
  style?: never;
  variant?: FormHelperTextVariants;
}

const FormHelperText: React.FC<FormHelperTextProps> = ({marginTop, children, variant, ...props}) => {
  let icon = null;
  switch (variant) {
    case FormHelperTextVariants.ERROR:
      icon = <ErrorIcon iconColor="colorTextError" decorative />;
      break;
    default:
      break;
  }

  return (
    <Flex vAlignContent="center" marginTop={marginTop || 'space20'}>
      {icon}
      <Text
        {...safelySpreadTextProps(props)}
        as="div"
        fontSize="fontSize30"
        lineHeight="lineHeight30"
        textColor={variant === 'error' ? 'colorTextError' : 'colorTextWeak'}
        marginLeft={icon ? 'space10' : undefined}
      >
        {children}
      </Text>
    </Flex>
  );
};

FormHelperText.displayName = 'FormHelperText';

FormHelperText.propTypes = {
  marginTop: PropTypes.oneOf(['space0']),
};

export {FormHelperText};
