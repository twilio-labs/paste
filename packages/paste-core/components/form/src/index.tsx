export * from './FormLabel';
export * from './FormInput';
export * from './FormHelpText';
export * from './FormTextArea';
export * from '@twilio-paste/checkbox';
export * from '@twilio-paste/radio-group';
export * from '@twilio-paste/select';
