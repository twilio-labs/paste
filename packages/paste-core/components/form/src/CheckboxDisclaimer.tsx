import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import {Checkbox, CheckboxProps} from './Checkbox';
import {FormHelpText} from './FormHelpText';

export interface CheckboxDisclaimerProps
  extends Omit<CheckboxProps, 'isSelectAll' | 'isSelectAllChild' | 'indeterminate' | 'helpText' | 'hasError'> {
  children: NonNullable<React.ReactNode>;
  errorText?: string | React.ReactNode;
}

const CheckboxDisclaimer: React.FC<CheckboxDisclaimerProps> = ({children, errorText, ...props}) => {
  return (
    <>
      <Box backgroundColor="colorBackground" borderRadius="borderRadius20" padding="space50">
        <Checkbox {...props} hasError={errorText != null}>
          {children}
        </Checkbox>
      </Box>
      {errorText && (
        <Box marginTop="space30" marginLeft="space50">
          <FormHelpText variant="error">{errorText}</FormHelpText>
        </Box>
      )}
    </>
  );
};

CheckboxDisclaimer.displayName = 'CheckboxDisclaimer';

if (process.env.NODE_ENV === 'development') {
  CheckboxDisclaimer.propTypes = {
    children: PropTypes.node.isRequired,
    errorText: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  };
}

export {CheckboxDisclaimer};
