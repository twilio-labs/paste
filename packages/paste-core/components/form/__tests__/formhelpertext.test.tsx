import * as React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {Theme} from '@twilio-paste/theme';
import {FormHelperText} from '../src';

describe('FormHelperText render', () => {
  it('it should render', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <FormHelperText>child</FormHelperText>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('it should render with an error icon', (): void => {
    const tree = renderer
      .create(
        <Theme.Provider theme="console">
          <FormHelperText variant="error">child</FormHelperText>
        </Theme.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('FormHelperText variant prop', () => {
  const initialProps = {
    variant: 'error',
  };

  const container = shallow(<FormHelperText {...initialProps} />);

  it('should have an error icon', () => {
    expect(container.find('ErrorIcon').length).toEqual(1);
  });

  it('should have colorTextError', () => {
    expect(container.find('Text').prop('textColor')).toEqual('colorTextError');
  });
});

describe('FormHelperText marginTop prop', () => {
  const initialProps = {
    marginTop: 'space0',
  };

  const container = shallow(<FormHelperText {...initialProps} />);

  it('should have marginTop: space0', () => {
    expect(container.find('Flex').prop('marginTop')).toEqual('space0');
  });
});
