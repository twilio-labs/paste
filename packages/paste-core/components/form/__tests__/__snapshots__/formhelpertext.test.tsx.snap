// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormHelperText render it should render 1`] = `
.emotion-2 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-2 *,
.emotion-2 *::after,
.emotion-2 *::before {
  box-sizing: border-box;
}

.emotion-1 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 0.25rem;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #666a6d;
}

<div
  className="emotion-2"
>
  <div
    className="emotion-1"
    display="flex"
  >
    <div
      className="emotion-0"
      fontSize="fontSize30"
    >
      child
    </div>
  </div>
</div>
`;

exports[`FormHelperText render it should render with an error icon 1`] = `
.emotion-4 {
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #282a2b;
  font-weight: 400;
}

.emotion-4 *,
.emotion-4 *::after,
.emotion-4 *::before {
  box-sizing: border-box;
}

.emotion-3 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 0.25rem;
}

.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: 1.5rem;
  height: 1.5rem;
  color: #e22525;
}

.emotion-2 {
  margin: 0;
  padding: 0;
  font-size: 0.875rem;
  line-height: 1.5rem;
  color: #e22525;
  margin-left: 0.125rem;
}

<div
  className="emotion-4"
>
  <div
    className="emotion-3"
    display="flex"
  >
    <span
      className="emotion-0 emotion-1"
      size="sizeIcon30"
    >
      <svg
        aria-hidden={true}
        aria-labelledby="1"
        height="100%"
        role="img"
        viewBox="0 0 24 24"
        width="100%"
      >
        <path
          d="M12 2a9.968 9.968 0 017.06 2.918l.011.01.011.012a9.967 9.967 0 012.913 6.765L22 12c0 5.523-4.477 10-10 10a9.969 9.969 0 01-7.066-2.924l-.005-.005A9.973 9.973 0 012 12C2 6.477 6.477 2 12 2zm6.709 4L5.999 18.707a8.966 8.966 0 006 2.293c4.883 0 8.856-3.886 8.997-8.72l.004-.266-.004-.28A8.96 8.96 0 0018.709 6zm-6.71-3A9 9 0 003 12c0 2.305.867 4.408 2.292 6L18 5.293A8.97 8.97 0 0012 3z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
    </span>
    <div
      className="emotion-2"
      fontSize="fontSize30"
    >
      child
    </div>
  </div>
</div>
`;
